codex-rs/core/src/git_info.rs

```diff
@@ -97,6 +124,249 @@ async fn run_git_command_with_timeout(args: &[&str], cwd: &Path) -> Option<std::
+                    for line in text.lines() {
+                        let line = line.trim();
+                        if let Some(rest) = line.strip_prefix("HEAD branch:") {
+                            let name = rest.trim();
+                            if !name.is_empty() {
+                                return Some(name.to_string());
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    // No remote-derived default; try common local defaults if they exist
+    for candidate in ["main", "master"] {
+        if let Some(verify) = run_git_command_with_timeout(
+            &[
+                "rev-parse",
+                "--verify",
+                "--quiet",
+                &format!("refs/heads/{candidate}"),
+            ],
+            cwd,
+        )
+        .await
+        {
+            if verify.status.success() {
+                return Some(candidate.to_string());
+            }
+        }
+    }
+
+    None
+}
+
+/// Build an ancestry of branches starting at the current branch and ending at the
+/// repository's default branch (if determinable)..
+async fn branch_ancestry(cwd: &Path) -> Option<Vec<String>> {
+    // Discover current branch (ignore detached HEAD by treating it as None)
+    let current_branch = run_git_command_with_timeout(&["rev-parse", "--abbrev-ref", "HEAD"], cwd)
+        .await
+        .and_then(|o| {
+            if o.status.success() {
+                String::from_utf8(o.stdout).ok()
+            } else {
+                None
+            }
+        })
+        .map(|s| s.trim().to_string())
+        .filter(|s| s != "HEAD");
+
+    // Discover default branch
+    let default_branch = get_default_branch(cwd).await;
+
+    let mut ancestry: Vec<String> = Vec::new();
+    if let Some(cb) = current_branch.clone() {
+        ancestry.push(cb);
+    }
+    if let Some(db) = default_branch {
+        if ancestry.first().map(|b| b != &db).unwrap_or(true) {
+            ancestry.push(db);
+        }
+    }
+
+    // Ensure we return Some vector, even if empty, to allow caller logic to proceed
+    Some(ancestry)
+}
+
+async fn find_closest_sha(cwd: &Path, branches: &[String], remotes: &[String]) -> Option<String> {
+    // A sha and how many commits away from HEAD it is.
+    let mut closest_sha: Option<(String, usize)> = None;
+    'branch_loop: for branch in branches {
+        for remote in remotes {
+            let remote_ref = format!("refs/remotes/{remote}/{branch}");
+            let verify_output = match run_git_command_with_timeout(
+                &["rev-parse", "--verify", "--quiet", &remote_ref],
+                cwd,
+            )
+            .await
+            {
+                Some(output) => output,
+                None => continue 'branch_loop,
+            };
+            if !verify_output.status.success() {
+                continue;
+            }
```

Comment:
As a general practice, I put a blank line after an early return. This function is pretty dense (no blank lines at all), so I think it would be a good move in this case.